dialogue-core/src/main/java/com/palantir/dialogue/core/CautiousIncreaseAggressiveDecreaseConcurrencyLimiter.java

```diff
@@ -122,36 +126,52 @@ void dropped() {
          */
         void success() {
             inFlight.decrementAndGet();
-            int newLimit = limit.accumulateAndGet(inFlightSnapshot, LimitUpdater.SUCCESS);
+            double newLimit = accumulateAndGetLimit(inFlightSnapshot, LimitUpdater.SUCCESS);
             if (log.isDebugEnabled()) {
                 log.debug("Increasing limit {}", SafeArg.of("newLimit", newLimit));
             }
         }
     }
 
-    enum LimitUpdater implements IntBinaryOperator {
+    enum LimitUpdater implements DoubleBinaryOperator {
         SUCCESS() {
             @Override
-            public int applyAsInt(int originalLimit, int inFlightSnapshot) {
-                if (inFlightSnapshot * 2 >= originalLimit) {
-                    return Math.min(MAX_LIMIT, originalLimit + 1);
+            public double applyAsDouble(double originalLimit, double inFlightSnapshot) {
+                if (inFlightSnapshot >= originalLimit * BACKOFF_RATIO) {
+                    // The limit is raised more easily when the maximum limit is low, and becomes linearly more
+                    // stubborn as the limit increases. Given a fixed rate of traffic this should result in
+                    // linear slope as opposed to the exponential slope expected from a static increment
+                    // value.
+                    double increment = 1D / originalLimit;
+                    return Math.min(MAX_LIMIT, originalLimit + increment);
                 }
                 return originalLimit;
             }
         },
         DROPPED() {
             @Override
-            public int applyAsInt(int originalLimit, int _inFlightSnapshot) {
-                return Math.max(MIN_LIMIT, (int) (originalLimit * BACKOFF_RATIO));
+            public double applyAsDouble(double originalLimit, double _inFlightSnapshot) {
+                return Math.max(MIN_LIMIT, originalLimit * BACKOFF_RATIO);
             }
         };
     }
 
+    private double accumulateAndGetLimit(int value, DoubleBinaryOperator accumulatorFunction) {
+        AtomicDouble localLimit = limit;
```

Comment:
why save limit into another variable rather than accessing the field?

Reply:
under contention this can become a very hot loop, the jit doesn't actually assume that final fields are truly final in most cases due to potential reflection hackery (part of the motivation for jigsaw!) and we don't opt into the experimental `-XX:+TrustFinalNonStaticFields`.

Relevant post describing performance improvements to Atomic*FieldUpdaters which special-cased trusting final non-static fields:
https://shipilev.net/blog/2015/faster-atomic-fu/

Some context:
https://bugs.openjdk.java.net/browse/JDK-8132243
https://bugs.openjdk.java.net/browse/JDK-8058164

Reply:
I _knew_ there was gonna be some beautiful java perf reason under all this 😂. Could we capture that first sentence in a comment maybe?